aif360/algorithms/inprocessing/gerryfair_classifier.py

```diff
@@ -0,0 +1,275 @@
+        X, X_prime, y = clean.extract_df_from_ds(dataset)
+        learner = Learner(X, y, self.predictor)
+        auditor = Auditor(dataset, self.fairness_def)
+        history = ClassifierHistory()
+
+        # initialize variables
+        n = X.shape[0]
+        costs_0, costs_1, X_0 = auditor.initialize_costs(n)
+        metric_baseline = 0
+        predictions = [0.0] * n
+
+        # scaling variables for heatmap
+        vmin = None
+        vmax = None
+
+        # print output variables
+        errors = []
+        fairness_violations = []
+
+        iteration = 1
+        while iteration < self.max_iters:
+            # learner's best response: solve the CSC problem, get mixture decisions on X to update prediction probabilities
+            history.append_classifier(learner.best_response(costs_0, costs_1))
+            error, predictions = learner.generate_predictions(history.get_most_recent(), predictions, iteration)
+            # auditor's best response: find group, update costs
+            metric_baseline = auditor.get_baseline(y, predictions)
+            group = auditor.get_group(predictions, metric_baseline)
+            costs_0, costs_1 = auditor.update_costs(costs_0, costs_1, group, self.C, iteration, self.gamma)
+
+            # outputs
+            errors.append(error)
+            fairness_violations.append(group.weighted_disparity)
+            self.print_outputs(iteration, error, group)
+            vmin, vmax = self.save_heatmap(iteration, dataset, history.get_most_recent().predict(X), vmin, vmax)
+            iteration += 1
+
+            # early termination:
+            if early_termination and (len(errors) >= 5) and (
+                    (errors[-1] == errors[-2]) or fairness_violations[-1] == fairness_violations[-2]) and \
+                    fairness_violations[-1] < self.gamma:
+                iteration = self.max_iters
+
+        self.classifiers = history.classifiers
+        if return_values:
+            return errors, fairness_violations
+
+    def predict(self, dataset, threshold=.5):
+        """
+        Function to return dataset object where labels are the predictions returned by the fitted model.
+
+        :param dataset:
+        :param threshold: The positive prediction cutoff for the soft-classifier.
+        :return: modified dataset object
+        """
+
+        # Generates predictions. We do not yet advise using this in sensitive real-world settings.
+        dataset_new = copy.deepcopy(dataset)
+        data, _, _ = clean.extract_df_from_ds(dataset_new)
+        num_classifiers = len(self.classifiers)
+        y_hat = None
+        for c in self.classifiers:
+            new_predictions = np.multiply(1.0 / num_classifiers, c.predict(data))
+            if y_hat is None:
+                y_hat = new_predictions
+            else:
+                y_hat = np.add(y_hat, new_predictions)
+        dataset_new.labels = tuple([1 if y >= threshold else 0 for y in y_hat])
```

Comment:
not sure this correct @hoffmansc  - the classifier has been trained to predict 1's and 0's not favorable_label / unfavorable_label. If I just changed this, it would flip things around.. also FP and FN rate are defined explicitly with respect to 0's and 1's so doesn't make sense to talk about favorable/unfavorable.